### PR TITLE
feature #3052 and #3054 - edit profile and qr code viewer

### LINKED ISSUE #3052: New edit profile screen
### User Story

As a user I want to enjoy while I edit my Status profile on a beautiful new Edit Profile screen.

### Description

*Type*: Feature

*Summary*: 

Replace the old edit screen with the new profile screen but with editable name and profile image. Add all image change related actions and menus.

screens 3/4 and 5/6 from the left - https://zpl.io/VQo59qn

### LINKED ISSUE #3054: Review and where needed restyle remaining Profile related screens
[comment]: # (Please replace ... with your information. Remove < and >)

### User Story

As a user I want to view and interact with screens which look nice and consistent 

### Description

*Type*: Feature

*Summary*: Update the following screens:

- Select network
- Show QR code

https://zpl.io/VQo59qn

### PR DESCRIPTION
**Pull request by goranjovic**
fixes #3052 
fixes #3054

### Summary:

Added a QR code viewer as per redesign
Added edit mode for Profile screen

### Testing notes (optional):

Link to OS specific Notification settings screen is in scope of another issue - #3051 

### Steps to test:
- Open Status
- Go to Profile tab
- Open QR viewer, try sharing the public key
- Try editing name or image

status: ready

### CODE REVIEW

**Comment by goranjovic** [comment]
Android QR viewer:

screenshot-2018-01-23_10 43 21 529

Android edit mode:

screenshot-2018-01-23_10 43 05 779

iOS 6s QR viewer:

simulator screen shot jan 19 2018 at 12 56 34 pm

iOS 5s QR viewer:

simulator screen shot jan 19 2018 at 12 54 14 pm

File: src/status_im/ui/components/chat_icon/screen.cljs [hunk 1]
```diff
@@ -2,13 +2,14 @@
   (:require-macros [status-im.utils.views :refer [defview]])
   (:require [re-frame.core :refer [subscribe dispatch]]
             [status-im.ui.components.react :refer [view
-                                                text
-                                                image]]
+                                                   text
```
Comment (jeluard) [inline src/status_im/ui/components/chat_icon/screen.cljs]: Can use move to `:as` usage?

File: src/status_im/ui/components/qr_code/styles.cljs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+   :width            (:width dimensions)
+   :align-items      :center
+   :justify-content  :center
+   :padding          qr-code-padding
+   :border-radius    8})
+
+(def toolbar-done-text-ios
+  {:color          colors/blue
+   :letter-spacing -0.2
+   :font-size      17})
```
Comment (jeluard) [inline src/status_im/ui/components/qr_code/styles.cljs]: Those can most probably be constant.

File: src/status_im/ui/components/qr_code/styles.cljs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+(defstyle name-container
+  {:flex           0.7
+   :flex-direction :column
+   :ios            {:align-items :center}
+   :android        {:margin-left    15}})
```
Comment (jeluard) [inline src/status_im/ui/components/qr_code/styles.cljs]: Too many spaces?

File: src/status_im/ui/components/qr_code/styles.cljs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+(defstyle name-text
+  {:color     colors/black
+   :font-size 17
+   :ios       {:letter-spacing -0.2}})
+
+(def address-text
+  {:color     colors/white
+   :font-size 12})
```
Comment (jeluard) [inline src/status_im/ui/components/qr_code/styles.cljs]: Use a constant?

File: src/status_im/ui/components/qr_code/styles.cljs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+(def toolbar-action-container
+  {:flex            0.15
+   :flex-direction  :column
+   :align-items     :center
+   :justify-content :center})
+
+(def toolbar-action-icon-container
+  {:width           40
```
Comment (jeluard) [inline src/status_im/ui/components/qr_code/styles.cljs]: Icon sizes are available as constants in `styles` I believe.

File: src/status_im/ui/components/qr_code/styles.cljs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+(def hash-value-type
+  {:color          colors/black
+   :padding-bottom 5})
+
+(def hash-value-text
+  {:color             colors/black
+   :margin-horizontal 60
+   :text-align        :center
+   :font-size         15
+   :letter-spacing    -0.2
```
Comment (jeluard) [inline src/status_im/ui/components/qr_code/styles.cljs]: Use constants?

File: src/status_im/ui/components/qr_code/views.cljs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+                  :number-of-lines 1} name]]
+    [react/view styles/toolbar-action-container
+     [react/touchable-highlight {:on-press #(list-selection/open-share {:message qr-value})}
+      [react/view styles/toolbar-action-icon-container
+       [vi/icon :icons/share {:color :black}]]]]]])
+
+(defn qr-viewer-body [qr-value dimensions]
+  [react/view {:style     styles/qr-code
+               :on-layout #(let [layout (.. % -nativeEvent -layout)]
+                             ;;TODO(goranjovic): check if we really really need to do this
```
Comment (jeluard) [inline src/status_im/ui/components/qr_code/views.cljs]: Could we address that before the merge?

File: src/status_im/ui/screens/profile/events.cljs [hunk 8]
```diff
@@ -172,17 +171,26 @@
...
           {old-status :status} (get accounts current-account-id)
           {:keys [status photo-path]} (:my-profile/profile db)
           cleaned-name (clean-name db :my-profile/profile)
-          cleaned-edit {:name         cleaned-name
-                        :status       status
-                        :photo-path   photo-path
-                        :last-updated now}]
+          cleaned-edit (if photo-path
+                         {:name         cleaned-name
+                          :status       status
```
Comment (jeluard) [inline src/status_im/ui/screens/profile/events.cljs]: Can we do this without the redundancy?

File: src/status_im/ui/screens/profile/styles.cljs [hunk 9]
```diff
@@ -4,7 +4,7 @@
             [status-im.utils.platform :as platform])
   (:require-macros [status-im.utils.styles :refer [defstyle]]))
 
-(defstyle toolbar-edit-text
+(defstyle toolbar-ios-text
```
Comment (jeluard) [inline src/status_im/ui/screens/profile/styles.cljs]: Is is only for iOS?

File: src/status_im/ui/screens/profile/styles.cljs [hunk 10]
```diff
@@ -61,6 +61,13 @@
 (def context-menu-custom-styles
   {:optionsContainer {:margin-top 78}})
 
+(def edit-profile-toolbar-icon
+  {:width           40
```
Comment (jeluard) [inline src/status_im/ui/screens/profile/styles.cljs]: Use constants?

File: src/status_im/ui/screens/profile/styles.cljs [hunk 11]
```diff
@@ -75,10 +82,18 @@
    :android {:font-size 12}})
 
 (defstyle profile-name-text
-  {:margin-top 8
-   :font-size  15
+  {:padding-vertical 14
+   :font-size        15
+   :text-align       :center
+   :ios              {:letter-spacing -0.2}
```
Comment (jeluard) [inline src/status_im/ui/screens/profile/styles.cljs:60]: Use constants? Or maybe extract a component?

File: src/status_im/ui/screens/profile/styles.cljs [hunk 12]
```diff
@@ -75,10 +82,18 @@
...
+   :android          {:color colors/black}})
+
+(defstyle profile-name-input-text
+  {:font-size  15
    :text-align :center
-   :ios        {:letter-spacing -0.2}
+   :ios        {:letter-spacing -0.2
```
Comment (jeluard) [inline src/status_im/ui/screens/profile/styles.cljs:66]: Use constants?

File: src/status_im/ui/screens/profile/views.cljs [hunk 13]
```diff
@@ -22,19 +22,33 @@
...
 (defn my-profile-toolbar []
   [toolbar/toolbar {}
    nil
    [toolbar/content-title ""]
    [react/touchable-highlight
-    {:on-press #(re-frame/dispatch [:my-profile/edit-profile])}
+    {:on-press #(re-frame/dispatch [:my-profile/start-editing-profile])}
     [react/view
-     [react/text {:style      styles/toolbar-edit-text
+     [react/text {:style      styles/toolbar-ios-text
```
Comment (jeluard) [inline src/status_im/ui/screens/profile/views.cljs]: Why iOS?

File: src/status_im/ui/screens/profile/views.cljs [hunk 14]
```diff
@@ -22,19 +22,33 @@
...
+(defn my-profile-edit-toolbar []
+  [toolbar/toolbar {}
+   nil
+   [toolbar/content-title ""]
+   [react/touchable-highlight
+    {:on-press #(re-frame/dispatch [:my-profile/save-profile])}
+    (if platform/ios?
+      [react/view
+       [react/text {:style styles/toolbar-ios-text}
+        (i18n/label :t/done)]]
```
Comment (jeluard) [inline src/status_im/ui/screens/profile/views.cljs]: Good candidate for a helper function? We have those bars in several places.

File: src/status_im/ui/screens/profile/views.cljs [hunk 15]
```diff
@@ -66,6 +80,47 @@
...
+
+(defn profile-name-input [name]
+  [react/view
+   [react/text-input
+    {:style          styles/profile-name-input-text
+     :placeholder    ""
+     :default-value  name
+     :on-focus       #(re-frame/dispatch [:my-profile/edit-profile])
+     :on-change-text #(re-frame/dispatch [:my-profile/update-name %])
+     ;;:on-end-editing #(re-frame/dispatch [:my-profile/save-profile])
```
Comment (jeluard) [inline src/status_im/ui/screens/profile/views.cljs]: Cleanup.

File: src/status_im/ui/components/qr_code/views.cljs [hunk 16]
```diff
@@ -0,0 +1,63 @@
+(ns status-im.ui.components.qr-code.views
```
Comment (jeluard) [inline src/status_im/ui/components/qr_code/views.cljs]: Does it make sense to have `viewer` in the component name? We also have a qr code scanner component.

**Review by jeluard — CHANGES_REQUESTED** [review]

File: src/status_im/ui/components/qr_code/styles.cljs (hunk 2 above)
Comment (goranjovic) [reply src/status_im/ui/components/qr_code/styles.cljs]: If I do that, we'll end up with constants like `font-size-12`, `font-size-17` the same way we did with colors. In order to do this, we need a font size palette as well with names like `normal`, `small`, `large`, `title`, etc.

File: src/status_im/ui/screens/profile/styles.cljs (hunk 9 above)
Comment (goranjovic) [reply src/status_im/ui/screens/profile/styles.cljs]: Yes, in new designs android uses icons for actions in toolbars, while ios uses labels, e.g. checkmark vs Done

File: src/status_im/ui/components/chat_icon/screen.cljs (hunk 1 above)
Comment (goranjovic) [reply src/status_im/ui/components/chat_icon/screen.cljs]: I'm kind of careful not to do that to every file I touch, but sure why not.

File: src/status_im/ui/components/qr_code/views.cljs (hunk 16 above)
Comment (goranjovic) [reply src/status_im/ui/components/qr_code/views.cljs]: Good point!

File: src/status_im/ui/screens/profile/events.cljs [hunk 17]
```diff
@@ -82,8 +82,7 @@
                      (assoc-in [:my-profile/profile :edit-status?] edit-status?)
                      (update-in [:my-profile/profile]
                                 #(merge (select-keys (get-current-account db) db/account-profile-keys) %)))]
-      {:db        new-db
-       :dispatch [:navigate-to :edit-my-profile]})))
+      {:db        new-db})))
```
Comment (yenda) [inline src/status_im/ui/screens/profile/events.cljs]: too many spaces now

File: src/status_im/ui/screens/profile/events.cljs (hunk 8 above)
Comment (yenda) [reply src/status_im/ui/screens/profile/events.cljs]: (if photo-path (assoc cleaned-edit :photo-path photo-path) cleaned-edit)

File: src/status_im/ui/components/qr_code/views.cljs (hunk 7 above)
Comment (goranjovic) [reply src/status_im/ui/components/qr_code/views.cljs]: Apparently, we do need it. Removed the comment.

File: src/status_im/ui/screens/profile/styles.cljs (hunk 12 above)
Comment (goranjovic) [reply src/status_im/ui/screens/profile/styles.cljs:66]: Same thing as with colors or font sizes. I agree, but let's try to see what are the constant values and where they are used other than here.

File: src/status_im/translations/en.cljs [hunk 18]
```diff
@@ -183,6 +183,7 @@
    :choose-from-contacts                  "Choose from contacts"
    :no-contacts                           "No contacts yet"
    :show-qr                               "Show QR code"
+   :qr-code-public-key-hint               "Share one these codes\nto start chatting"
```
Comment (chadyj) [inline src/status_im/translations/en.cljs]: As discussed in Slack, change to `Share this code to start chatting`

**Comment by goranjovic** [comment]
@jeluard @yenda @chadyj pushed the changes

**Review by jeluard — APPROVED** [review]

**Review by yenda — APPROVED** [review]

**Comment by asemiankevich** [comment]
Android: https://i.diawi.com/ChjwZT
iOS: https://i.diawi.com/AUFVmk

1. When I capture or selecting picture from gallery - the changes are not visible in real time, only when I tap Done. Is that expected?
2. Missing line under avatar for iOS and also looks like the space between Done and user picture is too large

image
3. Long tap does not work is that expected too? Also Done button is not aligned properly (iPhone 7)
image
4. Too much space between QR and key
image
5. I am not sure what we have to do with the long names - its now impossible to see the beginning of a phrase when editing and when it is saved the reset is hidden under 3 dots but they are not clickable

image

6. Different drop-down for Android
image
7. No Badges on app icon for iOS

**Comment by goranjovic** [comment]
@asemiankevich 

1. Not sure, it should probably be visible immediately and saved after you tap done. 
2. Ok, see it.
3. Long tap on what, not sure I follow? Weird about "Done", I did already fix that for 5s. Would've expected that 7 has more screen.
4. That one looks broken on 5s if there is less space between.
5. Probably the only way to really fix this is to put a limit on username length.

**Comment by asemiankevich** [comment]
Branch: PR-3121
Android: https://i.diawi.com/tyHyxt
iOS: https://i.diawi.com/ESPdUP

1. Done button is fixed for iOS 
image
2. Autofocus added for both Android and iOS when tapping Edit being on Profile tab
image
3. Photo is updated immediately when changing
4. Missing line for Edit screen will be added as separate issue (bounty)
5. Drop down for android is fixed
image
6. Notifications are not shown when disabled in settings and are working when enabled.

I am moving this to merge

**Review by jeluard — APPROVED** [review]

**Review by flexsurfer — APPROVED** [review]

**Pull request merged**